### PR TITLE
Change conveyors to heartbeat (finisher) + Exceptions handling : Fixes #58, #961, #1411

### LINKED ISSUE #58: Change conveyors to heartbeat
_Migrated from JIRA [RUCIO-2761](https://its.cern.ch/jira/browse/RUCIO-2761)_

### PR DESCRIPTION
**Pull request by cserf**
Change conveyors to heartbeat (finisher) + Exceptions handling : Fixes #58, #961, #1411

### CODE REVIEW

**Comment by mlassnig** [comment]
build crashed, i restarted it

File: lib/rucio/transfertool/fts3.py [hunk 1]
```diff
@@ -12,25 +12,30 @@
...
+from requests.adapters import ReadTimeout
 from requests.packages.urllib3 import disable_warnings  # pylint: disable=import-error
 
 from dogpile.cache import make_region
 from dogpile.cache.api import NoValue
 
-from fts3.rest.client.easy import Context, delegate
-from fts3.rest.client.exceptions import BadEndpoint, ClientError, ServerError
+from fts3.rest.client.easy import Context, delegate  # pylint: disable=no-name-in-module,import-error
+from fts3.rest.client.exceptions import BadEndpoint, ClientError, ServerError, TransferToolTimeout, TransferToolWrongAnswer  # pylint: disable=no-name-in-module,import-error
```
Comment (bari12) [inline lib/rucio/transfertool/fts3.py]: These imports are wrong.
TransferToolTimeout, TransferToolWrongAnswer needs to be imported from rucio.common.exception

**Review by bari12 — CHANGES_REQUESTED** [review]

**Comment by bari12** [comment]
This changes the hashing from rule_id to id. Imho this is not right unless we also change the transfer grouping? This change is also not described in any of the linked issues?

**Comment by cserf** [comment]
I changed the transfer grouping to `dest_rse, activity`. But I agree the change can lead to potential issues. Another possibiliy is to define in `rucio.cfg` the variable we want to hash.

**Comment by cserf** [comment]
Up

File: lib/rucio/daemons/conveyor/common.py [hunk 2]
```diff
@@ -101,6 +104,9 @@ def submit_transfer(external_host, job, submitter='submitter', logging_prepend_s
         record_timer('daemons.conveyor.%s.submit_bulk_transfer.per_file' % submitter, (time.time() - start_time) * 1000 / len(job['files']))
         record_counter('daemons.conveyor.%s.submit_bulk_transfer' % submitter, len(job['files']))
         record_timer('daemons.conveyor.%s.submit_bulk_transfer.files' % submitter, len(job['files']))
+    except (TransferToolTimeout, TransferToolWrongAnswer) as error:
+        logging.error(prepend_str + str(error))
```
Comment (mlassnig) [inline lib/rucio/daemons/conveyor/common.py:108]: is it fine to `continue` here?

File: lib/rucio/transfertool/fts3.py [hunk 3]
```diff
@@ -174,8 +184,12 @@ def submit(self, files, job_params, timeout=None):
...
                                         timeout=timeout)
             record_timer('transfertool.fts3.submit_transfer.%s' % self.__extract_host(self.external_host), (time.time() - start_time) * 1000 / len(files))
-        except:
-            logging.warn('Could not submit transfer to %s - %s', self.external_host, str(traceback.format_exc()))
+        except ReadTimeout as error:
+            raise TransferToolTimeout(error)
+        except JSONDecodeError as error:
+            raise TransferToolWrongAnswer(error)
+        except Exception as error:
+            logging.warn('Could not submit transfer to %s - %s' % (self.external_host, str(error)))
```
Comment (mlassnig) [inline lib/rucio/transfertool/fts3.py:192]: same `continue` question

File: lib/rucio/daemons/conveyor/common.py (hunk 2 above)
Comment (bari12) [reply lib/rucio/daemons/conveyor/common.py:108]: I think it is, as the `eid` will be `None` and this is handled correctly later on and setting the request to `RequestState.SUBMISSION_FAILED`

File: lib/rucio/transfertool/fts3.py (hunk 3 above)
Comment (cserf) [reply lib/rucio/transfertool/fts3.py:192]: That's the old behaviour, so there shouldn't be any problem

**Review by bari12 — APPROVED** [review]

**Pull request merged**